packages/graphics/src/utils/buildLine.ts

```diff
@@ -94,129 +260,253 @@ function buildNonNativeLine(graphicsData: GraphicsData, graphicsGeometry: Graphi
+
+                verts.push(imx, imy);
+                verts.push(x1 + (perp1x * outerWeight), y1 + (perp1y * outerWeight));
+            }
+            else /* arc is inside */
+            {
+                verts.push(x1 - (perpx * innerWeight), y1 - (perpy * innerWeight));
+                verts.push(omx, omy);
+
+                indexCount += round(
+                    x1, y1,
+                    x1 - (perpx * innerWeight), y1 - (perpy * innerWeight),
+                    x1 - (perp1x * innerWeight), y1 - (perp1y * innerWeight),
+                    verts, false
+                ) + 4;
+
+                verts.push(x1 - (perp1x * innerWeight), y1 - (perp1y * innerWeight));
+                verts.push(omx, omy);
+            }
+        }
+        else
+        {
+            verts.push(imx, imy);
+            verts.push(omx, omy);
         }
     }
 
-    p1x = points[(length - 2) * 2];
-    p1y = points[((length - 2) * 2) + 1];
+    x0 = points[(length - 2) * 2];
+    y0 = points[((length - 2) * 2) + 1];
 
-    p2x = points[(length - 1) * 2];
-    p2y = points[((length - 1) * 2) + 1];
+    x1 = points[(length - 1) * 2];
+    y1 = points[((length - 1) * 2) + 1];
 
-    perpx = -(p1y - p2y);
-    perpy = p1x - p2x;
+    perpx = -(y0 - y1);
+    perpy = x0 - x1;
 
     dist = Math.sqrt((perpx * perpx) + (perpy * perpy));
     perpx /= dist;
     perpy /= dist;
     perpx *= width;
     perpy *= width;
 
-    verts.push(p2x - (perpx * r1), p2y - (perpy * r1));
+    verts.push(x1 - (perpx * innerWeight), y1 - (perpy * innerWeight));
+    verts.push(x1 + (perpx * outerWeight), y1 + (perpy * outerWeight));
 
-    verts.push(p2x + (perpx * r2), p2y + (perpy * r2));
+    if (!closedShape)
+    {
+        if (style.cap === LINE_CAP.ROUND)
+        {
+            indexCount += round(
+                x1 - (perpx * (innerWeight - outerWeight) * 0.5),
+                y1 - (perpy * (innerWeight - outerWeight) * 0.5),
+                x1 - (perpx * innerWeight),
+                y1 - (perpy * innerWeight),
+                x1 + (perpx * outerWeight),
+                y1 + (perpy * outerWeight),
+                verts,
+                false
+            ) + 2;
+        }
+        else if (style.cap === LINE_CAP.SQUARE)
+        {
+            indexCount += square(x1, y1, perpx, perpy, innerWeight, outerWeight, false, verts);
+        }
+    }
 
     const indices = graphicsGeometry.indices;
 
     // indices.push(indexStart);
-
-    for (let i = 0; i < indexCount - 2; ++i)
+    for (let i = indexStart; i < indexCount + indexStart - 2; ++i)
     {
-        indices.push(indexStart, indexStart + 1, indexStart + 2);
+        x0 = verts[(i * 2)];
+        y0 = verts[(i * 2) + 1];
+
+        x1 = verts[(i + 1) * 2];
+        y1 = verts[((i + 1) * 2) + 1];
+
+        x2 = verts[(i + 2) * 2];
+        y2 = verts[((i + 2) * 2) + 1];
+
+        /* Skip zero area triangles */
+        if ((x0 * (y1 - y2)) + (x1 * (y2 - y0)) + (x2 * (y0 - y1)) === 0)
```

Comment:
Lets not do strict equality in float arithmetic. 

Need eps2 here. 
If we use `Math.abs(x1-x2)>eps`. and `eps=1e-3`, then here we have to use `eps2 = eps * eps`

we can put eps in our graphicsutils. @eXponenta @bigtimebuddy can you guys help me a bit here?

Reply:
Huh? What do you need help with exactly @ivanpopelyshev?

Reply:
Is it fine to merge as it is and I'll adjust epsilon in separate PR?

I am sure that we need EPS there. Also we have to put that setting in GraphicsUtils

Reply:
If this is the only place, seems like an easy fix for @SukantPal. Otherwise, do it as a follow up PR later if you want to do something more extensive.

Reply:
I added a `settings.EPSILON` property. You can use that if you want to commit directly; otherwise, I'll resolve this tomorrow.

Reply:
ok, just remember that if for | x1 - x2 | you use eps, then for squares you have to use eps2. usually.